lock-api/src/main/java/com/palantir/lock/watch/ClientLockWatchEventLogImpl.java

```diff
@@ -0,0 +1,189 @@
+    private volatile Optional<IdentifiedVersion> latestVersion;
+
+    public static ClientLockWatchEventLogImpl create() {
+        return new ClientLockWatchEventLogImpl(ClientLockWatchSnapshotUpdaterImpl.create());
+    }
+
+    @VisibleForTesting
+    static ClientLockWatchEventLogImpl create(ClientLockWatchSnapshotUpdater snapshotUpdater) {
+        return new ClientLockWatchEventLogImpl(snapshotUpdater);
+    }
+
+    private ClientLockWatchEventLogImpl(ClientLockWatchSnapshotUpdater snapshotUpdater) {
+        this.snapshotUpdater = snapshotUpdater;
+        this.eventLog = new ConcurrentSkipListMap<>();
+        this.latestVersion = Optional.empty();
+    }
+
+    @Override
+    public synchronized Optional<IdentifiedVersion> processUpdate(
+            LockWatchStateUpdate update,
+            Optional<IdentifiedVersion> earliestVersion) {
+        ProcessingVisitor visitor;
+        if (newLeader(update)) {
+            visitor = new NewLeaderVisitor(earliestVersion);
+        } else {
+            visitor = new ProcessingVisitor(earliestVersion);
+        }
+        update.accept(visitor);
+        return latestVersion;
+    }
+
+    /**
+     * @param timestampToVersion mapping from timestamp to identified version from client-side event cache;
+     * @param version            latest version that the client knows about; should be before timestamps in the
+     *                           mapping;
+     * @return lock watch events that occurred from (exclusive) the provided version, up to (inclusive) the latest
+     * version in the timestamp to version map.
+     */
+    @Override
+    public synchronized TransactionsLockWatchEvents getEventsForTransactions(
+            Map<Long, IdentifiedVersion> timestampToVersion,
+            Optional<IdentifiedVersion> version) {
+        Preconditions.checkState(latestVersion.isPresent(), "Cannot get events when log does not know its version");
+        Optional<IdentifiedVersion> versionInclusive =
+                version.map(oldVersion -> IdentifiedVersion.of(oldVersion.id(), oldVersion.version() + 1));
+        IdentifiedVersion currentVersion = latestVersion.get();
+        /*
+        There are three cases to consider where we would return a snapshot:
+            1: they provide an empty version;
+            2. their version has a different UUID (i.e. refers to the wrong leader);
+            3. their version is behind our log.
+        Note that if their version is ahead of our log, or we do not have a version, an exception is thrown instead.
+         */
+        if (!versionInclusive.isPresent()
+                || !versionInclusive.get().id().equals(currentVersion.id())
+                || eventLog.floorKey(versionInclusive.get().version()) == null) {
+            return TransactionsLockWatchEvents.failure(snapshotUpdater.getSnapshot(currentVersion));
+        }
+
+        IdentifiedVersion toVersion = Collections.max(timestampToVersion.values());
```

Comment:
What happens when timestampToVersion is empty?

Reply:
According to `Collections.max`: `@throws NoSuchElementException if the collection is empty.`

I think throwing is the right behaviour (if there are no elements, then there will be no updates anyway, so that does not make sense), but I'll change this to use a Preconditions instead.